torch/_inductor/scheduler.py

```diff
@@ -4207,6 +4214,80 @@ def get_graph_partition_signature(
 
         return signatures[::-1]
 
+    def reorder_for_partition_based_on_devices(
+        self, nodes: list[BaseSchedulerNode]
+    ) -> list[BaseSchedulerNode]:
+        """
+        Given topologically sorted `nodes`, reorder by devices while
+        respecting dependencies. Specifically, when a node reads from
+        another device, execute all leading nodes on that device since
+        we already have a graph partition for that device.
+        """
+
+        result: list[BaseSchedulerNode] = []
+        device_to_nodes: dict[Optional[torch.device], list[BaseSchedulerNode]] = (
+            collections.defaultdict(list)
+        )
+
+        def visit(node: BaseSchedulerNode) -> None:
+            node_device = node.get_device()
+            for dep in node.unmet_dependencies:
+                if (
+                    buffer := self.name_to_buf.get(dep.name, None)
+                ) and buffer.get_device() != node_device:
+                    input_device = buffer.get_device()
+                    # the dependent node may have been processed
+                    if input_device_nodes := device_to_nodes.pop(input_device, None):
+                        result.extend(input_device_nodes)
+            device_to_nodes[node_device].append(node)
+
+        for node in nodes:
+            visit(node)
+
+        for on_device_nodes in device_to_nodes.values():
+            result.extend(on_device_nodes)
+
+        return result
+
+    def reorder_for_partition_with_simple_dependency(
+        self, nodes: list[BaseSchedulerNode]
+    ) -> list[BaseSchedulerNode]:
+        """
+        Reorder a node if it should be partitioned and has simple dependency:
+        1. move a partitioned node to the front if it has no dependency
+        2. move a partitioned node to the back if it is only used by OutputNode
+        3. otherwise do not reorder
+        """
+
+        front: list[BaseSchedulerNode] = []
+        middle: list[BaseSchedulerNode] = []
+        back: list[BaseSchedulerNode] = []
+
+        def only_output_user(node: BaseSchedulerNode) -> bool:
+            for buf in node.get_outputs():
```

Comment:
One other thing, semi relevant again to vllm.. 

If we have an extern op that is not cudagraphable in the middle of the graph and takes in an output, we can allocate the output inside the prior cudagraph region to avoid copy.